### ISSUE TITLE
Remove "get_" prefixes from function names

### ISSUE THREAD

**Issue by jgabry** [stan-dev/rstan#276]
Suggestion from @bob-carpenter:

>  I'd recommend removing "get_" prefixes;  "set_" is OK, but "get_" doesn't
>  really add any info, just four more characters.

I'm fine with deprecating the functions with `get_` prefixes (`get_sampler_params`, `get_inits`, etc.). I don't think we would have any name conflicts with existing functions, so we should be able to just drop the `get_` prefixes and keep the resulting function names. Eventually we can disallow the `get_` versions entirely.

**Comment by bgoodri** [comment]
This is not unrelated to the question of whether to move to reference classes for RStan3.

**Comment by bob-carpenter** [comment]
Do they dicatate the names of getters/setters?  If so,
I'd prefer to favor consistency over conciseness.
- Bob

**Comment by bgoodri** [comment]
By default, you can do $foo to get foo, but they also support $getfoo().

**Comment by sakrejda** [comment]
I've done a few packages using reference classes and I usually use the same
name for the getter/setter.  If you keep a protected internal variable
(e.g.-foo$bar__, then the getter setter foo$bar(x=NULL) can return the
value foo$bar__ when x is NULL and set otherwise (the syntax foo$bar <- 3
calls foo$bar(x=3)).  If it's a read-only variable you can just test 'x'
and throw an error if !is.null(x).  There's probably a few other ways of
doing this but if you don't protect read-only variables it gets easy for
users to wreck the state of the object.

**Comment by bob-carpenter** [comment]
> On Mar 10, 2016, at 3:16 PM, Krzysztof Sakrejda notifications@github.com wrote:
> 
> I've done a few packages using reference classes and I usually use the same
> name for the getter/setter. If you keep a protected internal variable
> (e.g.-foo$bar__, then the getter setter foo$bar(x=NULL) can return the
> value foo$bar__ when x is NULL and set otherwise (the syntax foo$bar <- 3
> calls foo$bar(x=3)). 

I take it then the use would be

get value of bar: foo$bar()  
set bar to 3:     foo$bar(3)   

The full form foo$bar(x = 3) looks really confusing.

I find that pretty confusing, but I find the rest of R confusing,
too.

> If it's a read-only variable you can just test 'x'
> and throw an error if !is.null(x). There's probably a few other ways of
> doing this but if you don't protect read-only variables it gets easy for
> users to wreck the state of the object.

Absolutely.  On construction, objects should be in a good
state and setters should maintain that good state.  That's why
I don't like the lm() return in R --- it's easy to set it into
an inconsistent state.
- Bob

**Comment by ariddell** [comment]
PyStan can mirror this change. It happens to be more in keeping with Python conventions.

**Comment by davharris** [comment]
Regarding reference semantics, I haven't felt much need for it in rstan. If rstan does go that route, it might be worth considering [R6 classes](https://cran.r-project.org/web/packages/R6/vignettes/Introduction.html) (designed and maintained by the RStudio folks) over reference classes:

> They’re similar to R’s built-in reference classes, but are simpler, smaller, and faster, and they allow inheritance across packages.

**Comment by bgoodri** [comment]
Yeah, R6 is on the table, although my sense is that John Chambers fixed
most of the limitations of reference classes that originally motivated the
RStudio people to create R6.
